### PR TITLE
[#175897626] Allow return same OTP if valid

### PR DESCRIPTION
**Pull request by AleDore**
#### List of Changes
- Add fiscalCode-OTP key pair store on Redis while generating a new OTP
- Allow return the same OTP if it still valid, avoiding generate a new one
- Add tests

#### Motivation and Context
Since a generated OTP is valid for a time-boxed period, we want to returns the same OTP if the user request for a new generation, and the previous generated OTP is still valid (aka Redis hits a key value pair with provided fiscalCode). See this [story](https://www.pivotaltracker.com/story/show/175897626) for further informations.

#### How Has This Been Tested?
Unit and integration tests done.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by AleDore** [comment]
Do you think we can merge @BurnedMarshal ?

File: GenerateOtp/__tests__/redis.test.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      )
+      .run();
+  });
+});
+
+describe("retrieveOtpByFiscalCode", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+  it("should return an error when if error occurs while retrieving fiscalCode", async () => {
```
Comment (BurnedMarshal) [inline GenerateOtp/__tests__/redis.test.ts:102]:
```suggestion
  it("should return an error when fiscalCode retrieve fail", async () => {
```

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: GenerateOtp/__tests__/redis.test.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        () => fail(),
+        _ => expect(isNone(_)).toBeTruthy()
+      )
+      .run();
+  });
+
+  it("should return an error when if error occurs while retrieving related fiscalCode's OTP", async () => {
+    getTaskMock.mockImplementationOnce(() => taskEither.of(some(anOtpCode)));
+    getTaskMock.mockImplementationOnce(() =>
+      fromLeft("Cannot retrieve OTP code")
```
Comment (BurnedMarshal) [inline GenerateOtp/__tests__/redis.test.ts:125]:
```suggestion
      fromLeft(new Error("Cannot retrieve OTP code"))
```

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: utils/config.ts [hunk 3]
```diff
@@ -53,15 +56,15 @@ export const IConfig = t.intersection([
...
   ).getOrElse(600 as NonNegativeInteger),
   REDIS_CLUSTER_ENABLED: fromNullable(process.env.REDIS_CLUSTER_ENABLED)
     .map(_ => _.toLowerCase() === "true")
     .toUndefined(),
-  REDIS_TLS_ENABLED: fromNullable(process.env.REDIS_TLS_ENABLED).map(
-    _ => _.toLowerCase() === "true"
-  ),
+  REDIS_TLS_ENABLED: fromNullable(process.env.REDIS_TLS_ENABLED)
+    .map(_ => _.toLowerCase() === "true")
+    .toUndefined(),
```
[the remark this answers is not in the document]
Comment (AleDore) [reply utils/config.ts:67]: Documented in README :)

File: GenerateOtp/redis.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    `${OTP_PREFIX}${otpCode}`,
+    JSON.stringify(payload),
+    otpTtl
+  )
+    .chain(
+      fromPredicate(
+        _ => _,
+        () => new Error("Cannot Store OTP")
+      )
+    )
```
Comment (BurnedMarshal) [inline GenerateOtp/redis.ts:47]: We should use more strict typings here. 
```suggestion
): TaskEither<Error, true> =>
  setWithExpirationTask(
    redisClient,
    `${OTP_PREFIX}${otpCode}`,
    JSON.stringify(payload),
    otpTtl
  )
    .chain(
      fromPredicate(
        (_: boolean): _ is true => _,
        () => new Error("Cannot Store OTP")
      )
    )
```
We need to change even the return type of `setWithExpirationTask` as `TaskEither<Error, true>`

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: GenerateOtp/redis.ts (hunk 4 above)
Comment (BurnedMarshal) [reply GenerateOtp/redis.ts:47]: Sorry. I see now that if we change the return type of `setWithExpirationTask` we can remove the `chain` with the `fromPredicate`

Comment (AleDore) [reply GenerateOtp/redis.ts]: Done in [94ca09e](https://github.com/pagopa/io-functions-cgn/pull/35/commits/94ca09e6e229acfe3c8dbd59cbe76c0e52c21b7b). Note that `_ => _` is correcty inferred to `true`

Comment (AleDore) [reply GenerateOtp/redis.ts]: >Sorry. I see now that if we change the return type of setWithExpirationTask we can remove the chain with the fromPredicate. YEP You're right :)

**Review by BurnedMarshal — APPROVED** [review]
LGTM

**Pull request merged**